File: c/ray.c
#include "ray.h"

#include <math.h>

#include "tuple.h"

ray newRay(tuple origin, tuple direction) {
    ray r = {origin, direction};
    return r;
};

tuple project(ray r, float distance) {
    return addTuple(r.origin, multiplyTuple(r.direction, distance));
}

int intersect(ray r, sphere s, float *out) {
    tuple sphereToRay = substractTuple(r.origin, newPoint(0, 0, 0));

    float a = dot(r.direction, r.direction);
    float b = 2 * dot(r.direction, sphereToRay);
    float c = dot(sphereToRay, sphereToRay) - 1;

    float discriminant = b * b - 4 * a * c;

    if (discriminant < 0) {
        return 0;
    }

    out[0] = (-1 * b - sqrt(discriminant)) / (2 * a);
    out[1] = (-1 * b + sqrt(discriminant)) / (2 * a);
    return 2;
}

File: go/render/ray.go
package render

type Ray struct {
	origin    Tuple
	direction Tuple
}

func NewRay(origin, direction Tuple) *Ray {
	return &Ray{
		origin:    origin,
		direction: direction,
	}
}

func (r *Ray) Project(distance float64) Tuple {
	return r.direction.Mul(distance).Add(r.origin)
}

func (r *Ray) Transform(m *Matrix) *Ray {
	return &Ray{
		origin:    m.MultiplyTuple(r.origin),
		direction: m.MultiplyTuple(r.direction),
	}
}
